src/librustc_resolve/lib.rs

```diff
@@ -1666,13 +1666,22 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
         let mut extern_prelude: FxHashSet<Name> =
             session.opts.externs.iter().map(|kv| Symbol::intern(kv.0)).collect();
 
-        // HACK(eddyb) this ignore the `no_{core,std}` attributes.
-        // FIXME(eddyb) warn (elsewhere) if core/std is used with `no_{core,std}`.
-        // if !attr::contains_name(&krate.attrs, "no_core") {
-        // if !attr::contains_name(&krate.attrs, "no_std") {
-        extern_prelude.insert(Symbol::intern("core"));
-        extern_prelude.insert(Symbol::intern("std"));
-        extern_prelude.insert(Symbol::intern("meta"));
+        if session.rust_2018() {
```

Comment:
Why are we not doing the logic in 2015 in 2018 as well?

Reply:
1) In 2018 crates outside of extern prelude are in disadvantage (they have to use `crate::my_crate`), so this would make life harder for `#[no_std]` people optionally using `std` under a feature (https://github.com/rust-lang/rust/issues/53166#issuecomment-425219286), unless something https://github.com/rust-lang/rust/pull/54658-like is implemented.
2) For 2018 we still have enough time, while for 2015 extern prelude is already stabilized on beta.

Reply:
I'd happily do this for 2018 as well despite `1)` if it were urgent, since it's the most conservative variant.

Reply:
Re. 2) -- the current beta becomes stable on 2018-10-26 so we should have time to wait with merging this until the next lang meeting (2018-10-04) right?

------------

Re. 1) it is my understanding that given `crate::my_crate`, then the path segment `my_crate` will always refer to an item defined or `use`d in the root of the current crate. Is that right?

It would be good if someone (not me) could write down all the rules for the module system in Rust 2015 and Rust 2018 somewhere so that it becomes a bit more easy to judge various changes; mainly it would be good since there have been so many small changes that it is hard reason about things.

As for people conditionally using `std` wouldn't they just write as @Nemo157 wrote in https://github.com/rust-lang/rust/pull/54658?:

```rust
#![cfg_attr(not(feature = "std"), no_std)]

use core::mem;

#[cfg(feature = "std")]
use std::boxed::Box;
```

Reply:
>Re. 2)

Sure.

>Re. 1) it is my understanding that given crate::my_crate, then the path segment my_crate will always refer to an item defined or used in the root of the current crate. Is that right?

Yes.

>As for people conditionally using std wouldn't they just write as @Nemo157 wrote in #54658?:

Yes, conditional `#[no_std]` would work if behavior on both edition is reset to what this PR does.
But perhaps it's impractical for some reasons? Need to ask people actually writing libraries.
As I understand, the motivation described in https://github.com/rust-lang/rust/issues/53166#issuecomment-425219286 came from a lang-team meeting, so perhaps you remember something from that?

(As I said, I'd personally be totally okay with applying this PR to both edition.)

(P.S. I have to leave now and will be available only in couple of days.)

Reply:
>  so perhaps you remember something from that?

Sorry, afraid not; but we'll know in a few days :)